src/cancellation_token.c

```diff
@@ -0,0 +1,212 @@
+        result = 0;
+    }
+
+    return result;
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN_REGISTRATION), cancellation_token_register_notify, THANDLE(CANCELLATION_TOKEN), cancellation_token, TCALL_DISPATCHER_TARGET_FUNC_TYPE_NAME(CANCELLATION_TOKEN_CANCEL_CALL), on_cancel, void*, context)
+{
+    THANDLE(CANCELLATION_TOKEN_REGISTRATION) result = NULL;
+
+    if (cancellation_token == NULL || on_cancel == NULL)
+    {
+        /*Codes_SRS_CANCELLATION_TOKEN_04_010: [ cancellation_token_register_notify shall fail and return NULL when cancellation_token is NULL. ]*/
+        /*Codes_SRS_CANCELLATION_TOKEN_04_011: [ cancellation_token_register_notify shall fail and return NULL when on_cancel is NULL. ]*/
+        LogError("Invalid args: THANDLE(CANCELLATION_TOKEN) cancellation_token = %p, TCALL_DISPATCHER_TARGET_FUNC_TYPE_NAME(CANCELLATION_TOKEN_CANCEL_CALL) on_cancel = %p, void* context = %p", cancellation_token, on_cancel, context);
+    }
+    else
+    {
+        CANCELLATION_TOKEN* token_ptr = THANDLE_GET_T(CANCELLATION_TOKEN)(cancellation_token);
+
+        /*Codes_SRS_CANCELLATION_TOKEN_04_013: [ cancellation_token_register_notify shall call TCALL_DISPATCHER_REGISTER_TARGET(CANCELLATION_TOKEN_CANCEL_CALL) to register on_cancel with the dispatcher. ]*/
+        TCALL_DISPATCHER_TARGET_HANDLE(CANCELLATION_TOKEN_CANCEL_CALL) cancel_notification_target_handle = TCALL_DISPATCHER_REGISTER_TARGET(CANCELLATION_TOKEN_CANCEL_CALL)(token_ptr->cancel_notification_call_dispatcher, on_cancel, context);
+        if (cancel_notification_target_handle == NULL)
+        {
+            /*Codes_SRS_CANCELLATION_TOKEN_04_012: [ cancellation_token_register_notify shall fail and return NULL when any underlying call fails. ]*/
+            LogError("TCALL_DISPATCHER_REGISTER_TARGET(CANCELLATION_TOKEN_CANCEL_CALL)(token_ptr->cancel_notification_call_dispatcher, on_cancel, context) failed.");
+        }
+        else
+        {
+            /*Codes_SRS_CANCELLATION_TOKEN_04_014: [ cancellation_token_register_notify shall call THANDLE_MALLOC(CANCELLATION_TOKEN_REGISTRATION) to allocate a CANCELLATION_TOKEN_REGISTRATION handle. ]*/
+            THANDLE(CANCELLATION_TOKEN_REGISTRATION) registration = THANDLE_MALLOC(CANCELLATION_TOKEN_REGISTRATION)(cancellation_token_registration_dispose);
+            if (registration == NULL)
+            {
+                /*Codes_SRS_CANCELLATION_TOKEN_04_012: [ cancellation_token_register_notify shall fail and return NULL when any underlying call fails. ]*/
+                LogError("THANDLE_MALLOC(CANCELLATION_TOKEN)(cancellation_token_dispose) failed.");
+            }
+            else
+            {
+                CANCELLATION_TOKEN_REGISTRATION* registration_ptr = THANDLE_GET_T(CANCELLATION_TOKEN_REGISTRATION)(registration);
+                registration_ptr->cancel_notification_target_handle = cancel_notification_target_handle;
```

Comment:
> registration_ptr->cancel_notification_target_handle = cancel_notification_target_handle; [](http://example.com/codeflow?start=15&length=89)

How did this compile?

TCALLs `should` be kind of THANDLEs, and the MAJOR advantage of THANDLE is that they are not assignable.

Reply:
Not sure. Looks like this is _not_ a `THANDLE`? `cancel_notification_target_handle` is a `TCALL_DISPATCHER_TARGET_HANDLE(CANCELLATION_TOKEN_CANCEL_CALL)` which expands to `TCALL_DISPATCHER_TARGET_HANDLE_CANCELLATION_TOKEN_CANCEL_CALL` which, according to VS intellisense, is this:

<img width="423" alt="image" src="https://github.com/Azure/c-util/assets/628128/69a2332e-707b-4ff4-8fe2-2b178a42e3a1">

And that looks like a regular old pointer to me.

Reply:
I shall leave this be since it doesn't seem to be a `THANDLE` to me. Further, in `TCALL_DISPATCHER_LL_UNREGISTER_TARGET` in `tcall_dispatcher_ll.h`, I see that it treats `call_target` as a regular pointer and calls `free` on it.